### ISSUE TITLE
JdkClientContainer RemoteEndpoint.Basic.sendText(String) send blocks indefinitely

### ISSUE THREAD

**Issue by nuessgens** [eclipse-ee4j/tyrus#678]
We're having Issues with the Client where a `javax.websocket.RemoteEndpoint.Basic.sendText(String)` will send indefinitely (/"deadlock"). The occurrence of this issue is very rare.

Stack trace of hanging (/blocked) thread:
```
sun.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
org.glassfish.tyrus.core.TyrusFuture.get(TyrusFuture.java:53)
org.glassfish.tyrus.core.TyrusRemoteEndpoint$Basic.processFuture(TyrusRemoteEndpoint.java:143)
org.glassfish.tyrus.core.TyrusRemoteEndpoint$Basic.sendText(TyrusRemoteEndpoint.java:80)
``` 

The same issue occurs with  `javax.websocket.RemoteEndpoint.Async.sendText(String, SendHandler)` were the onResult method is never called.

It seems that the request is actually really not sent.
After investigating further I recognized that the class `org.glassfish.tyrus.container.jdk.client.TaskQueueFilter` is *not* thread-safe. Or - to be precise - locked inconsistently.
As I understood the TaskQueueFilter should guarantee that only one write operation is executed at a time (no overlapping, no concurrent writes)

In case the write-method is called a write task is enqueued. If the caller was the first to enqueue a task (taskLock false -> true), processTask is called which will trigger the write to the downstreamFilter. Any subsequent write-task is enqueued; processTask is not executed (as there is already a task in progress; taskLock == true).
As the write-handler completed, the next task from the queue is processed and so on. After the queue is empty taskLock is _released_  (taskLock true-> false) so that the next write-method invocation would start described process again.

```java
    @Override
    void write(ByteBuffer data, CompletionHandler<ByteBuffer> completionHandler) {
        taskQueue.offer(new WriteTask(data, completionHandler));
        if (taskLock.compareAndSet(false, true)) {
            processTask();
        }
    }

    private void processTask() {
        final Task task = taskQueue.poll();
        if (task == null) {
            taskLock.set(false);
            return;
        }
        task.execute(this);
    }
```

*BUT* The _release_ criteria is not checked atomic.
Assume the following Scenario
# Thread-1 enters processTask(); `taskQueue.poll()` returns null
# Thread-2 invokes write; offers the write-task to the taskQueue
# Thread-2 checks if processTask() should be called to initiate task processing; `taskLock == true` so compareAndSet will return `false` and processTask *is not* invoked
# Thread-1 checks `task == null` returns true and taskLock is set to `false`

Now we have a queued task which will not be processed by any thread. It will _heal_ if another thread will call write. (processTask would catch-up the already queued tasks, but it may take some time). But in my scenario it will _never_heal because i call write from a single thread.

**Comment by vikinghawk** [comment]
I don't believe the RemoteEndpoint was intended to be thread-safe.
Check out the javadoc:
https://docs.oracle.com/javaee/7/api/javax/websocket/RemoteEndpoint.Basic.html

Specifically:
> If the websocket connection underlying this RemoteEndpoint is busy sending a message when a call is >made to send another one, for example if two threads attempt to call a send method concurrently, or if >a developer attempts to send a new message while in the middle of sending an existing one, the send >method called while the connection is already busy may throw an IllegalStateException.

I've found other implementations (such as Tomcat) to also not be thread safe. The doc could be better and state this more clearly and you could argue Tyrus should be throwing IllegalStateException in this scenario as well.

Regardless, this the approach i have taken in my code

`final RemoteEndpoint.Basic endpoint = session.getBasicRemote();
synchronized (endpoint) {
    endpoint.sendText("");
}
`

**Comment by nuessgens** [comment]
I know. Therefore sendText is only called by a single thread in my application (but the "deadlock"/ block is still occurring)

It's the internal Tyrus implementation that is not thread-safe. The TaskQueueFilter is accessed by two threads. My application thread when I invoke `RemoteEndpoint.Basic.sendText` **and** the Java NIO-Thread that calls the `CompletionHandler.completed` (or `CompletionHandler.failed`).

Ironically tyrus seems to have introduced the TaskQueueFilter _to allow_ (/support) concurrent access to Basic-/Async-Remote (that implies the queuing of the write-tasks). Without that intermediate-filter I would not run into that issue.

This code-block is called by the Java NIO-Thread (after send completed)
https://github.com/tyrus-project/tyrus/blob/89723e870c176a5e9a91c25ce743a9970ebd0012/containers/jdk-client/src/main/java/org/glassfish/tyrus/container/jdk/client/TaskQueueFilter.java#L158-L178

The more I think about it: Maybe the inconsistent locking is not the problem **but** the _order_ of completion invocation.

```java
        public void execute(final TaskQueueFilter queueFilter) {
            queueFilter.downstreamFilter.write(getData(), new CompletionHandler<ByteBuffer>() {
                //...
                @Override
                public void completed(ByteBuffer result) {
                    // ...
                    getCompletionHandler().completed(getData());
                    queueFilter.processTask();
                }
            });
        }
```
The target completionHandler is notified first (`getCompletionHandler().completed(getData())`) - so my application-thread waiting on the future may wake up -  **and after that** `processTask()` is triggered again.
Therefore application-thread tries to send next task while NIO-Thread is still going to the `processTask()`-routine.

I'll try to swap the lines and report if it solves the problem. One downside would be that a potentially queued followup task would be sent *before* the completionHandler of the previous task was notified. But - as @vikinghawk mentioned -  _there should not be_ any follow-up task as BasicRemote.sendText should be called mutually exclusive.

**Comment by vikinghawk** [comment]
Ok yes, my bad. I missed the part about you only sending from a single thread.

Good investigation on this. I have been using the grizzly version of tyrus for about 4 years now and have not seen a similar issue. I was considering switching to the jdk version as i no longer need to support JRE 1.6 but sounds like it would be better to stick with grizzly.

**Comment by nuessgens** [comment]
A little update:

swapping the lines at 
https://github.com/eclipse-ee4j/tyrus/blob/f6dfb4749019502c2030848dd48a19950543834c/containers/jdk-client/src/main/java/org/glassfish/tyrus/container/jdk/client/TaskQueueFilter.java#L151-L152
and 
https://github.com/eclipse-ee4j/tyrus/blob/f6dfb4749019502c2030848dd48a19950543834c/containers/jdk-client/src/main/java/org/glassfish/tyrus/container/jdk/client/TaskQueueFilter.java#L140-L141
fixes the problem for me. 

What do you think folks is the better fix?
* Make the class thread-safe due concurrency (which might occur between IO-Thread, and potential several application-threads) => better locking / atomicity of queue.poll and setting "running"-flag
* Or just fix that the TaskQueueFilter WriteTask CompletioanHandler process/notify-completed order. So that TaskQueueFilter does not cause multi-threading issues _per default_ 

With the second approach potential multithreaded (application-level) access may lead to blocked / hanging writes.

Or in general: Is _tyrus_ intended to be thread-safe and allow concurrent writes on Basic-/AsyncRemote?

As @vikinghawk  quoted the spec says 
> [...] send method called while the connection is already busy **may** throw an IllegalStateException

So either it should be supported **or** throw an IllegalStateException but never block forever ;-)

Depending on your opinions (fix option1 or option 2) I'd like to create a pull request.

**Comment by TechnoFred** [comment]
Hello, I believe I encountered this issue today while sending synchronous text messages to a remote endpoint from one thread.  My stack trace is below.  Is there still any interest in fixing this? 
 Interestingly, I let the deadlock sit and I went to bed.  When I awoke hours later the deadlock had cleared and the system was running again.

The spec seems vague with its use of the word MAY in the description, doesn't it?

"..MAY throw an IllegalStateException."

My vote is to throw an IllegalStateException whenever an attempt is made to send a message while the previous message has not finished sending.

Please let me know how I can help resolve this.

```
#94 prio=5 os_prio=0 tid=0x00007fdb3400a800 nid=0x477d waiting on condition [0x00007fdaf7ed3000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000f0e37200> (a java.util.concurrent.CountDownLatch$Sync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
	at org.glassfish.tyrus.core.TyrusFuture.get(TyrusFuture.java:53)
	at org.glassfish.tyrus.core.TyrusRemoteEndpoint$Basic.processFuture(TyrusRemoteEndpoint.java:143)
	at org.glassfish.tyrus.core.TyrusRemoteEndpoint$Basic.sendText(TyrusRemoteEndpoint.java:80)
```

**Comment by nuessgens** [comment]
I'll try to create a pull request any time soon (currently I'm a little busy but maybe I manage to create one in the next two weeks)

I'll try to introduce a unit-test to reproduce the issue and than fix the inconsistent locking so that the client would be thread safe (and theoretically could also be accessed multi-threaded by client-applications).

I think - like the bug currently is occurring on my observations - throwing an IllegalStateException is not an option. The client is internally not thread-safe. It is used and accessed single-threaded by my application. So a client-application would get an IllegalStateException  at random times without doing anything wrong. Of course that would be better than a _deadlock_ but also not a satisfying behavior.

**Comment by TechnoFred** [comment]
Thank you.  As soon as I can I will look more closely at my exact situation here to try and better understand the problem.  A thread that becomes lost in the mechanism and gets stuck waiting forever is absolutely unacceptable.  I'll update you here if I make any progress or come up with a different angle on the fix.  Thank you.

**Comment by TechnoFred** [comment]
UPDATE:  In my case the seemingly stuck thread is actually not stuck at all.  My application was sending synchronous text messages to web browser client websocket connections on mobile devices.  I noticed that in certain circumstances the client browser ( opera or chrome running on Android ) seems to be put into a 'pause' or 'sleep' state by the OS when not in the foreground for a period of time.  I'm not sure what to call it because I haven't done any Android programming.  When the client browser is in this state, these messages are cached by either the browser or the underlying system until a buffer apparently fills up.  When this buffer fills up, the websocket connection remains open but some mechanism on the client ( websocket or TCP ) does not allow the reception of any further messages.  The message that seems to be stuck forever is waiting on the backlog of messages on the client to clear.  In my case when I bring the client browser back into focus the backlog of messages on the client is processed and the channel is cleared for further message transmission.  Voila, my 'lost' thread on the server wakes up and continues sending the message that it was 'stuck' sending.   I expected the connection to die in this case but it does not appear to do so.  I haven't dug into it any deeper, but for me it doesn't appear to be a Tyrus problem.  This behavior seems correct to me now that I understand it.  In my case I will implement a mechanism in my own nested messaging protocol to avoid this client caching of messages.

**Comment by TechnoFred** [comment]
UPDATE: In my case the seemingly stuck thread is actually not stuck at all. My application was sending synchronous text messages to web browser client websocket connections on mobile devices. I noticed that in certain circumstances the client browser ( opera or chrome running on Android ) seems to be put into a 'pause' or 'sleep' state by the OS when not in the foreground for a period of time. I'm not sure what to call it because I haven't done any Android programming. When the client browser is in this state, messages are cached somewhere in the underlying system ( client or server side I'm not sure yet ) until a buffer apparently fills up. When this buffer fills up, the websocket connection remains open but some mechanism disallows the sending of further messages.  The thread on the server that seems to be stuck forever is waiting on the backlog of messages on the client to clear.  In my case when I bring the client browser back into focus the backlog of messages on the client is processed and the channel is cleared for further message transmission.  Voila, my 'lost' thread on the server wakes up and continues sending the message that it was 'stuck' sending.  I expected the websocket connection to die in this case but it does not appear to do so. I haven't dug into it any deeper, but for me this may not indicate a Tyrus problem at all.  It may be designed behavior.  This behavior seems to be somewhat correct to me now that I kind of understand it. I would like to understand where the caching of these messages is taking place.  Maybe I have control of this cache?  I suspect it's on the client, but I'm not sure yet.  In my case I might implement a mechanism in my own nested messaging protocol to avoid this caching of messages.
